### PR TITLE
Forward merging of trivial-reduction dims in producers

### LINKED ISSUE #1539: Replay with domains merged with trivial reductions
`TransformPropagator` fails in the fusion below:

```
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(1);
  fusion.addInput(tv0);

  auto tv1 = broadcast(tv0, {true, false});
  auto tv2 = sum(tv1, {0});
  auto tv3 = set(tv2);
  fusion.addOutput(tv3);

  tv2->merge(0);
  tv2->split(0, 4);

  TransformPropagator::from(tv2);
```
```
Inputs:
  T0_g[ iS0{i1} ], float
Outputs:
  T3_g[ iS5{i1} ], float

%kernel_math {
T1_l[ bS1{1}, iS2{i1} ] = broadcast( T0_g[ iS0{i1} ] )
T2_l[ rS3{1}, iS4{i1} ] = reduction( T1_l[ bS1{1}, iS2{i1} ], op = add, initial value = double(0), fused = 0 )
T3_g[ iS5{i1} ]
   = T2_l[ rS3{1}, iS4{i1} ];
}
```

```
C++ exception with description "it != replay_CasP.getReplay().end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/transform_replay.cpp":546, please report a bug to PyTorch. TransformReplay::replayCasP: Could not find producer CA axis, iS7{( ceilDiv(( 1 * i1 ), 4) )}, requested in replay. Producer: T2_l[ iS7{( ceilDiv(( 1 * i1 ), 4) )}, iS8{4} ]. Producer CA position: 2. Consumer: T3_g[ iS5{i1} ]
Exception raised from replayCasP at ../torch/csrc/jit/codegen/cuda/transform_replay.cpp:546 (most recent call first):
```

The problem is the merge of `T2`, which results in a non-reduction domain. `replayCasP` with `T3` as a consumer and `T2` as a producer fails as it can't replay the merge output domain, which is because it's composed by a reduction domain. However, since the output domain of this merge is not a reduction domain, it's not filtered out at this part of `replayCasP`:

https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/transform_iter.cpp#L649-L653
```
 // producer ids we need to match in consumer
  std::vector<IterDomain*> producer_CA_ids(
      producer->domain()->domain().begin(),
      producer->domain()->domain().begin() + producer_compute_at_axis);
  producer_CA_ids = TensorDomain::noReductions(producer_CA_ids);
```

So, one way to "fix" this issue is to exclude such domains derived from trivial reduction domains.

Alternatively, it may also make sense to process trivial reductions in a similar way as broadcast forwarding. In the above case, the transformations of `tv2:

```
  tv2->merge(0);
  tv2->split(0, 4);
```

We could forward the merge with the trivial reduction domain and map the output domain of `tv2->merge()` with `tv3->axis(0)`. 

The latter approach seems to be more generic.

For the meantime, I'll do the first approach.

### PR DESCRIPTION
**Pull request by naoyam**
Forwards merging of trivial reductions with non-reduction domains. Mostly just replicated the logic for the broadcast forwarding. Unique changes are marked as PR comments.

TODO
- [x] Clean up tests.

Fixes #1539

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 1]
```diff
@@ -631,90 +634,72 @@ struct ConsumerForwardingInfo {
...
+      auto inner = merge->inner();
+      auto outer = merge->outer();
+      if ((inner->isTrivialReduction() && !outer->isReduction()) ||
+          (outer->isTrivialReduction() && !inner->isReduction())) {
+        auto compliment_id = inner->isTrivialReduction() ? inner : outer;
+        auto forwarded_id = inner->isTrivialReduction() ? outer : inner;
+        forwarding_map.emplace(std::make_pair(forwarded_id, merge->out()));
+        compliment_map.emplace(std::make_pair(
+            forwarded_id, std::vector<IterDomain*>{compliment_id}));
+      }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp]: Forward trivial reduction merges

File: torch/csrc/jit/codegen/cuda/transform_iter.h [hunk 2]
```diff
@@ -159,7 +159,8 @@ class TORCH_CUDA_CU_API ReplayTransformations : public IterVisitor {
 class TORCH_CUDA_CU_API BestEffortReplay {
  private:
   std::unordered_map<IterDomain*, IterDomain*> target2replay_id_map_;
-  std::unordered_map<IterDomain*, IterDomain*> forward_id_map_;
+  std::unordered_map<IterDomain*, IterDomain*> replay_forward_id_map_;
+  std::unordered_map<IterDomain*, IterDomain*> target_forward_id_map_;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_iter.h:163]: Separate forwarding maps for targets and replays.

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 3]
```diff
@@ -261,18 +263,17 @@ BestEffortReplay::BestEffortReplay(
...
-            ir_utils::filterByType<IterDomain>(replay_expr->inputs());
-        replay_rfactor_ids.insert(inp_ids.begin(), inp_ids.end());
-      }
+    // Only want to forward rfactor in map
+    auto out_ids = ir_utils::filterByType<IterDomain>(replay_expr->outputs());
+    if (std::any_of(out_ids.begin(), out_ids.end(), [](IterDomain* id) {
+          return id->isRFactorProduct();
+        })) {
+      auto inp_ids = ir_utils::filterByType<IterDomain>(replay_expr->inputs());
+      replay_rfactor_ids.insert(inp_ids.begin(), inp_ids.end());
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp:276]: Moved out of the innermost loop as it doesn't depend on the loop of `replay_expr->inputs()`.

File: test/cpp/jit/test_gpu.cpp [hunk 4]
```diff
@@ -21452,6 +21452,200 @@ TEST_F(NVFuserTest, FusionVectorizeContigIndexWithBroadcast_CUDA) {
...
+  fusion.addOutput(tv3);
+
+  fusion.printMath();
+  // Merging a trivial reduction with a non-reduction domain
+  tv2->merge(0, 1);
+  tv2->split(0, 4);
+
+  tv3->split(0, 4);
+
+  fusion.printMath();
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: I assume you were just using this test for debugging?

Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: All tests need some print cleanup

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 5]
```diff
@@ -631,90 +634,72 @@ struct ConsumerForwardingInfo {
...
-
-  // producer ids we need to match in consumer
-  std::vector<IterDomain*> producer_CA_ids(
-      producer->domain()->domain().begin(),
-      producer->domain()->domain().begin() + producer_compute_at_axis);
-  producer_CA_ids = TensorDomain::noReductions(producer_CA_ids);
-
-  // If producer has an rfactor root, that's what will match the consumer
-  std::vector<IterDomain*> producer_root = producer->getMaybeRFactorDomain();
+struct ProducerForwardingInfo {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp:653]: Could you add a comment to this similar to the consumer variant?

**Review by csarofeen — APPROVED** [review]
LGTM just some minor cleanup before merging. Thanks!

**Comment by naoyam** [comment]
@csarofeen I changed this PR to allow forwarding only when trivial-reduction domains are root domains. Non-root trivial reduction domains are those created with split-by-one of reductions domains, for example:

```
tv0: [R0, I1]
tv0->split(0, 1); // the inner output is a trivial reduction domain
tv0->merge(1, 2); // merging with the trivial reduction domain
```

I think forwarding in patterns like this can be confusing. At the root domain level, there's nothing that can be forwarded, so allowing this forwarding would need some more changes in `TransformReplay::replayPasC/CasP`. Unless there's a compelling case where we would need to support forwarding of intermediate trivial reduction domains, I think it makes more sense to make this PR simpler.

**Pull request merged**